File: jjazz-pure-core/src/main/java/org/jjazz/purecore/harmony/Note.java
package org.jjazz.purecore.harmony;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;
import java.text.ParseException;
import java.util.Objects;

/**
 * A musical note with pitch, duration, velocity, and accidental display.
 * Immutable. Extracted from JJazzLab Note, stripped of ResUtil/NetBeans deps.
 */
public class Note implements Comparable<Note>, Cloneable {

    public static final int VELOCITY_MIN = 0;
    public static final int VELOCITY_STD = 100;
    public static final int VELOCITY_MAX = 127;
    public static final int PITCH_MIN = 0;
    public static final int PITCH_STD = 60; // C4 (middle C)
    public static final int PITCH_MAX = 127;
    public static final int OCTAVE_MIN = 0;
    public static final int OCTAVE_STD = 4;
    public static final int OCTAVE_MAX = 10;

    public enum Accidental { FLAT, SHARP }

    public static final String[] NOTES_FLAT = {"C","Db","D","Eb","E","F","Gb","G","Ab","A","Bb","B"};
    public static final String[] NOTES_SHARP = {"C","C#","D","D#","E","F","F#","G","G#","A","A#","B"};

    private final int pitch;
    private final SymbolicDuration symbolicDuration;
    private final float beatDuration;
    private final Accidental accidental;
    private final int velocity;
    private String pianoOctaveString; // cached

    public Note() {
        this(PITCH_STD);
    }

    public Note(int p) {
        this(p, SymbolicDuration.QUARTER, VELOCITY_STD, Accidental.FLAT);
    }

    public Note(int p, float bd) {
        this(p, bd, VELOCITY_STD);
    }

    public Note(int p, float bd, int v) {
        this(p, bd, v, Accidental.FLAT);
    }

    public Note(int p, SymbolicDuration sd, int v, Accidental alt) {
        Objects.requireNonNull(sd);
        Objects.requireNonNull(alt);
        checkArgument(checkPitch(p), "p=%s", p);
        checkArgument(checkVelocity(v), "v=%s", v);
        this.pitch = p;
        this.beatDuration = sd.getDuration();
        this.symbolicDuration = sd;
        this.accidental = alt;
        this.velocity = v;
    }

    public Note(int p, float bd, int v, Accidental alt) {
        Objects.requireNonNull(alt);
        checkArgument(checkPitch(p), "p=%s", p);
        checkArgument(bd > 0, "bd=%s", bd);
        checkArgument(checkVelocity(v), "v=%s", v);
        this.pitch = p;
        this.beatDuration = bd;
        this.symbolicDuration = SymbolicDuration.getSymbolicDuration(bd);
        this.accidental = alt;
        this.velocity = v;
    }

    public Note(Note n, int newPitch) {
        this(newPitch, n.beatDuration, n.velocity, n.accidental);
    }

    public Note(Note n, Accidental alt) {
        this(n.pitch, n.beatDuration, n.velocity, alt);
    }

    /**
     * Parse a note string like "C", "Db", "A#", "C!4", "Eb!3".
     * Octave range is [0-10].
     */
    public Note(String s) throws ParseException {
        Objects.requireNonNull(s);
        String str = s.trim();
        Accidental alt = Accidental.FLAT;

        if (str.isEmpty()) {
            throw new ParseException("Empty note string", 0);
        }

        String degreeStr = str.substring(0, 1);
        if (str.length() > 1 && (str.charAt(1) == 'b' || str.charAt(1) == '#')) {
            degreeStr = str.substring(0, 2);
        }

        String octaveStr = null;
        int octaveIndex = str.indexOf("!");
        if (octaveIndex == str.length() - 1) {
            throw new ParseException("Invalid note: " + str, str.length() - 1);
        }
        if (octaveIndex != -1) {
            octaveStr = str.substring(octaveIndex + 1);
        }

        int relPitch = -1;
        // Handle special cases first
        if (degreeStr.compareTo("Cb") == 0)      { relPitch = 11; alt = Accidental.FLAT; }
        else if (degreeStr.compareToIgnoreCase("B#") == 0) { relPitch = 0; alt = Accidental.SHARP; }
        else if (degreeStr.compareToIgnoreCase("E#") == 0) { relPitch = 5; alt = Accidental.SHARP; }
        else if (degreeStr.compareToIgnoreCase("Fb") == 0) { relPitch = 4; alt = Accidental.FLAT; }
        else {
            for (int i = 0; i < NOTES_FLAT.length; i++) {
                if (degreeStr.compareToIgnoreCase(NOTES_FLAT[i]) == 0) { relPitch = i; alt = Accidental.FLAT; break; }
                if (degreeStr.compareToIgnoreCase(NOTES_SHARP[i]) == 0) { relPitch = i; alt = Accidental.SHARP; break; }
            }
        }

        if (relPitch == -1) {
            throw new ParseException("Invalid note: " + str, 0);
        }

        int octave = OCTAVE_STD;
        if (octaveStr != null) {
            try {
                octave = Integer.parseInt(octaveStr);
            } catch (NumberFormatException e) {
                throw new ParseException("Invalid note: " + str + " : " + e.getLocalizedMessage(), 0);
            }
        }
        if (!checkOctave(octave)) {
            throw new ParseException("Invalid note: " + str, 0);
        }

        this.pitch = octave * 12 + relPitch;
        this.beatDuration = SymbolicDuration.QUARTER.getDuration();
        this.symbolicDuration = SymbolicDuration.QUARTER;
        this.accidental = alt;
        this.velocity = VELOCITY_STD;
    }

    @Override
    public Note clone() {
        return new Note(this.pitch, this.beatDuration, this.velocity, this.accidental);
    }

    public int getPitch() { return pitch; }
    public int getRelativePitch() { return pitch % 12; }
    public float getDurationInBeats() { return beatDuration; }
    public SymbolicDuration getSymbolicDuration() { return symbolicDuration; }
    public int getOctave() { return pitch / 12; }
    public Accidental getAccidental() { return accidental; }

    public int getRelativeAscInterval(Note relNote) {
        int delta = relNote.getRelativePitch() - getRelativePitch();
        return delta < 0 ? delta + 12 : delta;
    }

    public int getRelativeDescInterval(Note relNote) {
        int delta = getRelativePitch() - relNote.getRelativePitch();
        return delta < 0 ? delta + 12 : delta;
    }

    /**
     * Shortest pitch delta from this note's relative pitch to relPitch. [-5, +6].
     */
    public int getRelativePitchDelta(int relPitch) {
        checkArgument(relPitch >= 0 && relPitch <= 11);
        int pitchDelta = relPitch - getRelativePitch();
        if (pitchDelta > 6) pitchDelta -= 12;
        else if (pitchDelta < -5) pitchDelta += 12;
        return pitchDelta;
    }

    public Note getTransposed(int t) {
        return new Note(this, pitch + t);
    }

    public Note getCentered(int lowPitch, int highPitch) {
        checkArgument(lowPitch <= highPitch - 12);
        int newPitch = pitch;
        while (newPitch < lowPitch) newPitch += 12;
        while (newPitch > highPitch) newPitch -= 12;
        return new Note(this, newPitch);
    }

    public Note getTransposed(int pitchShift, int pitchLimit) {
        checkArgument(pitchLimit >= 13 && pitchLimit <= 119);
        int newPitch = this.pitch + pitchShift;
        if (pitchShift > 0) while (newPitch > pitchLimit) newPitch -= 12;
        else if (pitchShift < 0) while (newPitch < pitchLimit) newPitch += 12;
        return new Note(this, newPitch);
    }

    public Note getTransposedWithinOctave(int t) {
        int rp = getRelativePitch() + t;
        if (t != 0 && (t % 12) != 0) {
            rp = rp < 0 ? rp + 12 : rp;
            rp = rp > 11 ? rp - 12 : rp;
            rp = getOctave() * 12 + rp;
        }
        return new Note(this, rp);
    }

    public int getLowerPitch(int relPitch, boolean acceptEquals) {
        checkArgument(relPitch >= 0 && relPitch <= 11);
        int p = getOctave() * 12 + relPitch;
        if ((relPitch == getRelativePitch() && !acceptEquals) || relPitch > getRelativePitch()) {
            p = (getOctave() - 1) * 12 + relPitch;
        }
        return Math.max(p, 0);
    }

    public int getUpperPitch(int relPitch, boolean inclusive) {
        checkArgument(relPitch >= 0 && relPitch <= 11);
        int p = getOctave() * 12 + relPitch;
        if ((relPitch == getRelativePitch() && !inclusive) || relPitch < getRelativePitch()) {
            p = (getOctave() + 1) * 12 + relPitch;
        }
        return Math.min(p, 127);
    }

    public int getClosestPitch(int relPitch) {
        int up = getUpperPitch(relPitch, true);
        int low = getLowerPitch(relPitch, true);
        return (up - getPitch() > getPitch() - low) ? low : up;
    }

    public int getVelocity() { return velocity; }

    public boolean equalsRelativePitch(Note n) {
        return getRelativePitch() == n.getRelativePitch();
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (!(obj instanceof Note other)) return false;
        return this.pitch == other.pitch
            && Float.floatToIntBits(this.beatDuration) == Float.floatToIntBits(other.beatDuration)
            && this.velocity == other.velocity;
    }

    @Override
    public int hashCode() {
        int hash = 7;
        hash = 67 * hash + this.pitch;
        hash = 67 * hash + Float.floatToIntBits(this.beatDuration);
        hash = 67 * hash + this.velocity;
        return hash;
    }

    @Override
    public int compareTo(Note n) {
        int res = Integer.compare(pitch, n.pitch);
        if (res == 0) {
            res = Float.compare(beatDuration, n.beatDuration);
            if (res == 0) res = Integer.compare(velocity, n.velocity);
        }
        return res;
    }

    @Override
    public String toString() {
        return toPianoOctaveString();
    }

    public String toPianoOctaveBeatString() {
        return toPianoOctaveString() + ":" + beatDuration;
    }

    public String toRelativeNoteString() {
        return toRelativeNoteString(accidental);
    }

    public String toRelativeNoteString(Accidental acc) {
        return acc == Accidental.FLAT ? NOTES_FLAT[getRelativePitch()] : NOTES_SHARP[getRelativePitch()];
    }

    public String toPianoOctaveString() {
        if (pianoOctaveString == null) {
            pianoOctaveString = toRelativeNoteString() + (getOctave() - 1);
        }
        return pianoOctaveString;
    }

    public boolean isFlat() { return accidental == Accidental.FLAT; }

    public boolean isWhiteKey() { return isWhiteKey(this.pitch); }

    public int getWhiteKeyPitch() {
        int res = getPitch();
        if (!isWhiteKey()) {
            res = isFlat() ? res + 1 : res - 1;
            res = Math.max(0, Math.min(127, res));
        }
        return res;
    }

    // --- Static helpers ---

    public static boolean checkPitch(int p) { return p >= PITCH_MIN && p <= PITCH_MAX; }
    public static boolean checkVelocity(int v) { return v >= VELOCITY_MIN && v <= VELOCITY_MAX; }
    public static boolean checkOctave(int o) { return o >= OCTAVE_MIN && o <= OCTAVE_MAX; }

    public static boolean isWhiteKey(int pitch) {
        pitch = pitch % 12;
        return pitch != 1 && pitch != 3 && pitch != 6 && pitch != 8 && pitch != 10;
    }

    public static int getNormalizedRelPitch(int absPitch) {
        if (absPitch >= 0) return absPitch % 12;
        else return (12 - (-absPitch % 12)) % 12;
    }

    public static int limitPitch(int pitch, int lowPitch, int highPitch) {
        checkArgument(lowPitch <= highPitch - 11);
        int newPitch = pitch;
        while (newPitch < lowPitch) newPitch += 12;
        while (newPitch > highPitch) newPitch -= 12;
        return newPitch;
    }

    public static Note[] getChromaticNotesArray(int pitchFrom, int pitchTo) {
        checkArgument(pitchFrom >= 0 && pitchTo >= 0 && pitchFrom <= pitchTo);
        Note[] notes = new Note[pitchTo - pitchFrom + 1];
        for (int i = 0; i < notes.length; i++) notes[i] = new Note(pitchFrom + i);
        return notes;
    }

    public static Note parsePianoOctaveString(String s) throws ParseException {
        if (s == null || s.length() < 2) throw new ParseException("Invalid string s=" + s, 0);
        int index = (s.charAt(1) == '#' || s.charAt(1) == 'b') ? 2 : 1;
        String strNote = s.substring(0, index);
        String strOctave = s.substring(index);
        int octave = Integer.parseInt(strOctave) + 1;
        return new Note(strNote + "!" + octave);
    }

    public static Note parsePianoOctaveBeatString(String s) throws ParseException {
        Objects.requireNonNull(s);
        s = s.strip();
        int index = s.indexOf(":");
        if (index == -1) throw new ParseException("Invalid PianoOctaveBeat string: " + s, 0);
        Note n = parsePianoOctaveString(s.substring(0, index));
        String strDur = s.substring(index + 1);
        float dur;
        try { dur = Float.parseFloat(strDur); }
        catch (NumberFormatException ex) { throw new ParseException("Invalid duration: " + s, index + 1); }
        return new Note(n.getPitch(), dur);
    }

    public static String saveAsString(Note n, boolean skipAccidental) {
        checkNotNull(n);
        if (skipAccidental) return n.pitch + "," + n.velocity + "," + n.beatDuration;
        else return n.pitch + "," + n.accidental + "," + n.velocity + "," + n.beatDuration;
    }

    public static Note loadAsString(String s) throws ParseException {
        checkNotNull(s);
        String[] strs = s.split(",");
        if (strs.length == 4 || strs.length == 3) {
            try {
                int p = Integer.parseInt(strs[0]);
                Accidental alt = strs.length == 3 ? Accidental.FLAT : Accidental.valueOf(strs[1]);
                int v = Integer.parseInt(strs[strs.length == 3 ? 1 : 2]);
                float bd = Float.parseFloat(strs[strs.length == 3 ? 2 : 3]);
                return new Note(p, bd, v, alt);
            } catch (IllegalArgumentException ex) { /* fall through */ }
        }
        throw new ParseException("Note.loadAsString() Invalid Note string s=" + s, 0);
    }
}

File: jjazz-engine/src/harmony/note.rs
use std::fmt;

/// Musical note with pitch (0-127), duration in beats, velocity (0-127).
#[derive(Debug, Clone, Copy, PartialEq, serde::Serialize, serde::Deserialize)]
pub struct Note {
    pub pitch: u8,
    pub duration: f32,
    pub velocity: u8,
    pub accidental: Accidental,
}

#[derive(Debug, Clone, Copy, PartialEq, serde::Serialize, serde::Deserialize)]
pub enum Accidental { Flat, Sharp }

impl Default for Accidental { fn default() -> Self { Accidental::Flat } }

pub const PITCH_MIN: u8 = 0;
pub const PITCH_STD: u8 = 60;
pub const PITCH_MAX: u8 = 127;
pub const OCTAVE_STD: i32 = 4;
pub const VELOCITY_STD: u8 = 100;

pub const NOTES_FLAT: [&str; 12] = ["C","Db","D","Eb","E","F","Gb","G","Ab","A","Bb","B"];
pub const NOTES_SHARP: [&str; 12] = ["C","C#","D","D#","E","F","F#","G","G#","A","A#","B"];

impl Note {
    pub fn new(pitch: u8) -> Self {
        assert!(pitch <= PITCH_MAX);
        Self { pitch, duration: 1.0, velocity: VELOCITY_STD, accidental: Accidental::Flat }
    }

    pub fn with_dur_vel(pitch: u8, duration: f32, velocity: u8) -> Self {
        assert!(pitch <= PITCH_MAX && velocity <= 127 && duration > 0.0);
        Self { pitch, duration, velocity, accidental: Accidental::Flat }
    }

    pub fn with_all(pitch: u8, duration: f32, velocity: u8, accidental: Accidental) -> Self {
        assert!(pitch <= PITCH_MAX && velocity <= 127 && duration > 0.0);
        Self { pitch, duration, velocity, accidental }
    }

    pub fn relative_pitch(&self) -> u8 { self.pitch % 12 }

    pub fn octave(&self) -> i32 { self.pitch as i32 / 12 }

    pub fn rel_pitch_to_string(&self) -> String {
        match self.accidental {
            Accidental::Flat => NOTES_FLAT[self.relative_pitch() as usize].to_string(),
            Accidental::Sharp => NOTES_SHARP[self.relative_pitch() as usize].to_string(),
        }
    }

    pub fn piano_octave_string(&self) -> String {
        format!("{}{}", self.rel_pitch_to_string(), self.octave() - 1)
    }

    /// Transpose by `semitones` semitones.
    pub fn transpose(&self, semitones: i32) -> Self {
        let new_pitch = (self.pitch as i32 + semitones).clamp(0, 127) as u8;
        Self { pitch: new_pitch, ..*self }
    }

    /// Get the closest pitch matching `rel_pitch` to this note.
    pub fn closest_pitch(&self, rel_pitch: u8) -> u8 {
        let up = self.upper_pitch(rel_pitch, true);
        let low = self.lower_pitch(rel_pitch, true);
        if up - self.pitch > self.pitch - low { low } else { up }
    }

    pub fn lower_pitch(&self, rel_pitch: u8, inclusive: bool) -> u8 {
        let base = self.octave() * 12 + rel_pitch as i32;
        let mut p = base;
        let rp = self.relative_pitch();
        if (rp == rel_pitch && !inclusive) || rel_pitch > rp {
            p = (self.octave() - 1) * 12 + rel_pitch as i32;
        }
        p.max(0) as u8
    }

    pub fn upper_pitch(&self, rel_pitch: u8, inclusive: bool) -> u8 {
        let base = self.octave() * 12 + rel_pitch as i32;
        let mut p = base;
        let rp = self.relative_pitch();
        if (rp == rel_pitch && !inclusive) || rel_pitch < rp {
            p = (self.octave() + 1) * 12 + rel_pitch as i32;
        }
        p.min(127) as u8
    }

    pub fn equals_relative_pitch(&self, other: &Note) -> bool {
        self.relative_pitch() == other.relative_pitch()
    }
}

impl Default for Note {
    fn default() -> Self { Self::new(PITCH_STD) }
}

impl fmt::Display for Note {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.piano_octave_string())
    }
}

/// Parse a note from a string like "C", "Eb", "F#", "C!4", "Bb!3"
pub fn parse_note(s: &str) -> Result<Note, String> {
    let s = s.trim();
    if s.is_empty() { return Err("empty".into()); }

    let mut accidental = Accidental::Flat;
    let (degree_str, rest) = if s.len() >= 2 && (s.as_bytes()[1] == b'b' || s.as_bytes()[1] == b'#') {
        (&s[..2], &s[2..])
    } else {
        (&s[..1], &s[1..])
    };

    let mut rel_pitch: i32 = -1;
    if degree_str.eq_ignore_ascii_case("Cb") { rel_pitch = 11; accidental = Accidental::Flat; }
    else if degree_str.eq_ignore_ascii_case("B#") { rel_pitch = 0; accidental = Accidental::Sharp; }
    else if degree_str.eq_ignore_ascii_case("E#") { rel_pitch = 5; accidental = Accidental::Sharp; }
    else if degree_str.eq_ignore_ascii_case("Fb") { rel_pitch = 4; accidental = Accidental::Flat; }
    else {
        for i in 0..12 {
            if degree_str.eq_ignore_ascii_case(NOTES_FLAT[i]) { rel_pitch = i as i32; accidental = Accidental::Flat; break; }
            if degree_str.eq_ignore_ascii_case(NOTES_SHARP[i]) { rel_pitch = i as i32; accidental = Accidental::Sharp; break; }
        }
    }
    if rel_pitch < 0 { return Err(format!("invalid note: {}", s)); }

    let octave = if let Some(idx) = rest.find('!') {
        rest[idx+1..].parse::<i32>().map_err(|e| format!("bad octave: {}", e))?
    } else {
        OCTAVE_STD
    };

    let pitch = (octave * 12 + rel_pitch) as u8;
    if pitch > 127 { return Err(format!("pitch out of range: {}", pitch)); }

    Ok(Note { pitch, duration: 1.0, velocity: VELOCITY_STD, accidental })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_note_pitch() {
        let c4 = Note::new(60);
        assert_eq!(c4.pitch, 60);
        assert_eq!(c4.relative_pitch(), 0);
        assert_eq!(c4.octave(), 5);
        assert_eq!(c4.piano_octave_string(), "C4");
    }

    #[test]
    fn test_transpose() {
        let c4 = Note::new(60);
        assert_eq!(c4.transpose(2).pitch, 62);
    }

    #[test]
    fn test_closest_pitch() {
        let c4 = Note::new(60);
        assert_eq!(c4.closest_pitch(11), 59); // B3
        assert_eq!(c4.closest_pitch(1), 61);  // Db4
    }

    #[test]
    fn test_parse() {
        let n = parse_note("Eb!3").unwrap();
        assert_eq!(n.pitch, 39); // 3*12+3
        let n2 = parse_note("F#").unwrap();
        assert_eq!(n2.pitch, 54); // 4*12+6
    }
}
